src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -830,6 +830,11 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
 
```

Comment:
What about cancellation here?

Reply:
Similarly on the FlushOutgoingBytesAsync call below.

Reply:
Specifically on FlushOutgoingBytesAsync? I decided against allowing cancellation mid-frame, as it would force us to tear the whole connection down. If you meant cancellation between serializing the headers and writing them to the wire though, I can probably plumb in cancellation there.

Reply:
Yeah, that's a good point. We don't want to actually cancel the write to the stream.

That said, it seems that we can wait on FlushOutgoingBytesAsync indefinitely and cancellation will never kick in. That seems bad.

(It would be nice to add a test for this -- write enough data in the request body to fill the socket send queue and cause it to pend. Loopback server needs to send a large window update to allow the client to proceed, but also not process the incoming frames, so the send queue backs up. Once the send actually pends, cancel the token and verify that the operation is actually cancelled instead of hanging waiting for a send completion.)

I'm not quite sure what to do about this.

I think we need FlushOutgoingBytesAsync to be cancellable, but cancelling it will *not* cancel the stream write. Instead it will just return a cancelled task as soon as cancellation happens, but leave the connection write pending. We could do this by registering for the cancellation token and returning a cancelled Task; this gets a little messy but it seems doable.

However, I think this creates another issue. If FlushOutgoingBytesAsync is cancelled, we will end up releasing the connection writer lock, allowing someone else to take it. But at this point, the pending write may still not be complete. I think that means we need to track whether we have a pending send or not, and if so, await it when we try to take the writer lock. 

Thoughts?

Reply:
Just to be clear, I think this corresponds to steps (3) and (6) in your list above, correct?

Reply:
Yeah, this does correspond to steps (3) and (6).

Just to make sure we're on the same page: you're suggesting we minimize the time between when a client cancels the CancellationToken and when we complete the TaskCompletionSource with an exception.

We currently only handle FlushOutgoingBytesAsync as if it were atomic -- we'll cancel before or after it completes, but not during it. If we implement what you suggested, we'll be able to report a cancellation to whoever initiated the request during FlushOutGoingBytesAsync. We won't actually cancel the write though, and will still block the connection along with any other pending requests.

I think the implementation you suggested seems doable, if the benefits here outweigh the additional complexity.

Reply:
I wouldn't use the term "minimize". In general it's fine for cancellation to take some time to propagate. 

What's not okay, it seems to me, is for cancellation to be unable to happen because of external factors we have no control over. That seems to be the case here. If the server doesn't process even unblock our socket send queue, cancellation will never happen.

Reply:
Okay, I think that's reasonable. I'll try to get this working in the next iteration.

Reply:
This should be working in the latest commit, along with a new test that checks this behavior.